dev/com.ibm.ws.ejbcontainer/src/com/ibm/ws/ejbcontainer/osgi/internal/EJBRuntimeImpl.java

```diff
@@ -1683,6 +1686,26 @@ protected void unsetLibertyFeature(ServiceReference<LibertyFeature> feature) {
         }
     }
 
+    @Reference(cardinality = ReferenceCardinality.OPTIONAL, policy = ReferencePolicy.DYNAMIC)
+    protected void setCheckpointPhase(CheckpointPhase phase) {
+        checkpointPhase = phase;
+    }
+
+    protected void unsetCheckpointPhase(CheckpointPhase phase) {
+        checkpointPhase = null;
+    }
+
+    /**
+     * Determines if application start should optimize for checkpoint after all configured applications
+     * have started.
+     *
+     * @return true if application start should optimize for checkpoint; false otherwise.
+     */
+    @Override
+    public boolean isCheckpointApplications() {
+        return CheckpointPhase.APPLICATIONS == checkpointPhase;
```

Comment:
@tjwatson I'm not aware of any potential EJB optimizations for CRIU that could occur prior to application start.

For some bean types, EJB Container attempts to defer some processing, such as loading the bean classes until first use; the property optimized here, EJBInitAtStartup, specifically moves loading of customer classes to application start.

With or without this change, if any of this EJB runtime code is running during DEPLOYMENT, then there is a significant problem for EJB applications.  EJBInitAtStartup only a hint to defer; it is not applicable to all bean types and configurations, so there will definitely be application code run in this code path.

Exactly how can I tell which EJB runtime code might be executed during DEPLOYMENT?

Reply:
> Exactly how can I tell which EJB runtime code might be executed during DEPLOYMENT?

I'm not sure I understand this question or at least how to answer it because I do not know the EJB runtime code.  For DEPLOYMENT the startup path is going to be the same as APPLICATIONS except as soon as the runtime does any class static initialization of application code (run the app class `<clinit>`) a checkpoint will be done before proceeding to run the code in the application.  This ensures that if the DEPLOYMENT phase is used that we will not execute user application code prior to the checkpoint being taken.  Does that help?

If I understand your comment above then I think it is appropriate to only have this change for the APPLICATIONS phase regardless.

Reply:
Ah, I was thinking it was based on appmanager or runtime framework lifecycle; but since it will be based on a class initializer, I did some further testing of how this configuration option works.  At this point in time, enabling this option for CRIU will cause application code to run (a `<clinit>`) much earlier than would otherwise occur, thus reducing the benefit one would see with DEPLOYMENT.  To avoid that, at least for this first change, I'd prefer to stay with APPLICATIONS for now.

However, as it turns out, at least the main reason classes would be loaded earlier is a bug; which was identified by testing from @dettmoney, related to the second config option in issue #20368.  Once I deliver the fix to that bug, and then can enable pool pre-loading for CRIU, I should also be able to switch this change for `EJBInitAtStartup` to also apply to DEPLOYMENT in addition to APPLICATIONS.

Net is; this is a good catch, but enabling for DEPLOYMENT will need to be done later.  To that end, I've added isCheckpointDeployment to the EJBRuntime interface for use later.

Reply:
I'm fine with only doing this for APPLICATIONS for now.